### ISSUE TITLE
CNTK generated ONNX File fails to load

### ISSUE THREAD

**Issue by dmagee** [microsoft/onnxruntime#1685]
**Describe the bug**
The following onnx file fails to load in the C API:

https://drive.google.com/open?id=1mdLaDgBFWDYU-KNJ9zoToLk2FcMhDtMZ

The error is:

C:\Software\onnxruntime-win-x64-gpu-0.5.0\Examples-build\Release>example-app.exe
Using Onnxruntime C API
Load model from TISSUETYPE_8X_DSNET_12D.cntk2.7.onnx failed:[ShapeInferenceError] Attribute strides has incorrect size

This file was exported from CNTK under windows. It loads with netron (https://lutzroeder.github.io/netron/), and python onnx module, but not TensorRT (see https://devtalk.nvidia.com/default/topic/1061738/tensorrt/tensorrt-fails-to-load-onnx-file-created-by-cntk/post/5376384/#5376384).

**Urgency**
Stops people migrating away from CNTK.

**System information**
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows 10
- ONNX Runtime installed from (source or binary): Binary 
- ONNX Runtime version: 0.5.0
- Python version: n/a
- Visual Studio version (if applicable): 2017
- CUDA/cuDNN version: n/a ?
- GPU model and memory: NVIDIA 1080

**To Reproduce**
Use Slightly modified version of https://github.com/microsoft/onnxruntime/blob/master/csharp/test/Microsoft.ML.OnnxRuntime.EndToEndTests.Capi/C_Api_Sample.cpp with model name changed (also had to replace ORT_ENABLE_BASIC with the code 1 as it doesn't seem to be  defined in v0.5.0).

**Expected behavior**
Should load model and nor report error as above.

**Additional context**
I suspect this could be something to do with inferred dimensions/parameters. Some ONNX code (netron, python onnx module) deal with such things well, others (onnxruntime c api, TensorRT) do not. I'm not sure if CNTK is producing invalid onnx or if the libraries that don't work aren't flexible enough, or maybe there is an onnx  version mismatch.

**Comment by faxu** [comment]
@ebarsoumMS : can you see if this is an issue with cntk export?

**Comment by dmagee** [comment]
The same file actually has issues reading in CNTK, although a bugfix to the CNTK reader sorts that. Worth linking the issue here:

https://github.com/microsoft/CNTK/issues/3746

**Comment by dmagee** [comment]
@ebarsoumMS any progress on this? Is it likely a CNTK issue or a c++ onnxruntime issue? I had a quick look at the CNTK code and it seems to use onnx runtime (possibly not the latest version??) to do the export.

**Comment by huangyiping-ai** [comment]
I have the same problem. Have you solved it?

**Comment by huangyiping-ai** [comment]
@dmagee 
I converted the pytorch model into onnx model.
When I run
Import onnx
Model_onnx_path= "LPRNet_model.onnx"
Model = onnx. load (model_onnx_path)
Onnx. checker. check_model (model)
When onnx version = 1.1.? The shape error of onnx. checker. check_model
When onnx version = 1.3.0, onnx. checker. check_model passes
So, I guess on onnxruntime, is it also a version error? For reference only, the guess may be wrong.

**Comment by dmagee** [comment]
@huangyping-ai that looks like you are using the python version of onnxruntime. My model works with the python version, just not with the c++ version. I think you probably have a slightly different issue? I may be wrong.

**Comment by dmagee** [comment]
Update: I'm not sure if anyone at Microsoft is looking into this (@faxu, @ebarsoumMS ?), but I downloaded the  onnxruntime source. The C API seems to be in a state of flux (my example and one of the provided examples doesn't compile due to various missing functions), but I got a C++ example working and it gives the same error as the binary release with c API:

 Load model from TISSUETYPE_8X_DSNET_12D.cntk2.7.onnx failed:[ShapeInferenceError] Attribute strides has incorrect size

I tracked this down to old.cc ~ line 60:

```
  std::vector<int64_t> strides;
  if (getRepeatedAttribute(ctx, "strides", strides)) {
    if (strides.size() != n_input_dims) {
      fail_shape_inference("Attribute strides has incorrect size");
    }
  } else {
    strides.assign(n_input_dims, 1);
  }
```

strides.size() seems to be 2, whereas n_input_dims is 3. The latter is correct I think. The input is an RGB image so 3 x 1024 x 1024. All functions in the network deal with 3D data (w*h*rgb or w*h*nfmaps).

Still not sure how to fix it though. CNTK uses onnxruntime to write the file.

D.

**Comment by dmagee** [comment]
Further update: This seems to be an issue with pooling node (convPoolShapeInference1(), old.cc). This network does 2x2 max pooling on 3D data (W x H x RGB or later  W x H x nFmaps). To achieve this in CNTK you provide a 2x2 shape description (CNTK::NDShape(2, 2) in C++). Obviously this has an implicit 3rd dimension (the kernel is effectively 2x2x1). The same issue applies to strides and pad as well i.e. they are specified as 2D, but are implicitly 3D. It seems they are output to the onnx file in this implicit form. onnxruntime doesn't seem to handle this when reading the file. I tried adding a few lines to the parser (in old.cc) in the appropriate places between load and checks:

    while(strides.size() < n_input_dims) {
        strides.push_back(1) ;
    }
...
      while (kernel_shape.size() < n_input_dims) {
          kernel_shape.push_back(1);
      }
...
      while (pads.size() < n_input_dims*2) {
          pads.push_back(0);
      }

However, now I'm getting the error:

failed:Node:Pooling86 Output:Pooling86_Output_0 [ShapeInferenceError] Can't merge shape info. Both source and target dimension have values but they differ. Source=15 Target=32 Dimension=2

This pool should go from 1x32x1024x512 -> 1 x 32x511x255 (i.e. no padding in any dimension, and no pooling in the fmaps dimension), however my change above to the kernel shape seems to make it pool in that dimension as well causing the error (at least that's what I think is happening).

I still have no idea if this is an onnx issue or a cntk one (I guess that depends on what's in the onnx standard about implicit dimensions - is there even a standard??). However, netron and the python version of onnx seem to handle this fine.

**Comment by faxu** [comment]
@dmagee @huangyiping-ai are you still impacted by this issue with the latest version of ONNX Runtime? If so and you still require assistance, please let us know and we can investigate further.

**Comment by YanYas** [comment]
Hi @faxu , I've come across this problem too. I'll test with the latest version of Onnx and report back.

**Comment by YanYas** [comment]
```
[ErrorCode:Fail] Load model from C:\Users\Igor\Documents\VLML\PythonResearchProjects\PythonExportCNTKToOnnx\VAE_Model_4.onnx failed:[ShapeInferenceError] Attribute dilations has incorrect size
   at Microsoft.ML.OnnxRuntime.NativeApiStatus.VerifySuccess(IntPtr nativeStatus)
   at Microsoft.ML.OnnxRuntime.InferenceSession.Init(String modelPath, SessionOptions options)
   at Microsoft.ML.OnnxRuntime.InferenceSession..ctor(String modelPath, SessionOptions options)
   at VLML.RunnerCore.UpdateDEBUG(PatchTracer tracer, Path ModelFile, IEnumerable`1 Input_Data...
```
Same error for two different models (one a [VAE ](https://drive.google.com/file/d/1judW2Xw2JlKdJRt6JSWMv57PPRQ6_wJt/view?usp=sharing) the other a ResNet Transfer Learning example which I'll link to when its finished uploading.

Thanks for looking into this

**Comment by dmagee** [comment]
I thought I'd commented last week, but yes it is still an issue in the latest binary release.

**Comment by KeDengMS** [comment]
From ONNX [spec](https://github.com/onnx/onnx/blob/master/docs/Operators.md#MaxPool) and [implementation](https://github.com/onnx/onnx/blob/master/onnx/defs/nn/defs.cc#L138-L140), it seems to only support pooling2D for 4D input, and pooling3D for 5D input. The model exported from CNTK has a 5D input for pooling ops, but the kernel shape is 2D. In the same model, there are conv ops having reshape before/after to make sure input is 4D, and IMO Pooling ops should do the same in CNTK exporter.

For Netron and python onnx module, shape inference is either not invoked or silently ignored. Before the bug is fixed, a workaround is to edit the MaxPool/AveragePool nodes in the model, with Squeeze/Unsqueeze (assuming batch==1) before and after to make sure the input to pooling ops are 4D:
```
import onnx
from onnx import helper, shape_inference
mp = onnx.load('/path/to/model')
mp_fix = onnx.ModelProto()
mp_fix.CopyFrom(mp)
mp_fix.graph.ClearField('node')
mp_fix.graph.ClearFiled('value_info')
for n in mp.graph.node:
  if n.op_type in ['MaxPool', 'AveragePool']:
    mp_fix.graph.node.add().CopyFrom(helper.make_node('Squeeze', n.input, [n.input[0] + '_squeezed'], n.name + '_squeeze_before', axes=[1]))
    pool_node = mp_fix.graph.node.add()
    pool_node.CopyFrom(n)
    pool_node.input[0] += '_squeezed'
    pool_node.output[0] += '_before_unsqueeze'
    mp_fix.graph.node.add().CopyFrom(helper.make_node('Unsqueeze', pool_node.output, n.output, n.name + '_unsqueeze_after', axes=[1]))
  else:
    mp_fix.graph.node.add().CopyFrom(n)
mp_fix = shape_inference.infer_shapes(mp_fix)
onnx.save(mp_fix, '/path/to/fixed_model')
```

Now the fixed model should have no issue in loading in ONNX runtime.

**Comment by faxu** [comment]
@dmagee does the fixed model work for you? It seems the root cause of the issue will take some time to investigate in CNTK and is lower priority at this time. Let us know if you're blocked on this with some urgency.

**Comment by dmagee** [comment]
I'm afraid we abandoned trying to get onnxruntime to work. We're working in a commercial environment and the fact that you couldn't sort this properly (even given the fact that we practically gave you the solution above) in a sensible timeframe meant we went looking for other solutions, Our networks work with TensorRT now (as of v7) and we're going forward with that, which has the added advantage of native TensorFlow support when we eventually move away from CNTK. 

If you want to fix onnxruntime it for others I suggest looking at my comments above. I think the issue with the "Can't merge shape info." was that I added the implicit dimensions at the wrong end of the array. I suspect it's pretty easy to get onnxruntime to support this, and given both CNTK and pyTorch have generated such models possibly a good idea?
